src/rt/rwlock.rs

```diff
@@ -0,0 +1,269 @@
+use crate::rt::{
+    object::{self, Object},
+    thread, Access, Synchronize,
+};
+
+use std::collections::HashSet;
+use std::sync::atomic::Ordering::{Acquire, Release};
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub(crate) struct RwLock {
+    obj: Object,
+}
+
+#[derive(Debug, PartialEq)]
+enum RwLockType {
+    Read(HashSet<thread::Id>),
+    Write(thread::Id),
+}
+
+#[derive(Debug)]
+pub(super) struct State {
+    /// If the rwlock should establish sequential consistency.
+    seq_cst: bool,
+
+    /// `Some` when the rwlock is in the locked state.
+    /// The `thread::Id` references the thread that currently holds the rwlock.
+    lock: Option<RwLockType>,
+
+    /// Tracks write access to the rwlock.
+    last_access: Option<Access>,
+
+    /// Causality transfers between threads
+    synchronize: Synchronize,
+}
+
+impl RwLock {
+    /// Common RwLock function
+    pub(crate) fn new(seq_cst: bool) -> RwLock {
+        super::execution(|execution| {
+            let obj = execution.objects.insert_rwlock(State {
+                seq_cst,
+                lock: None,
+                last_access: None,
+                synchronize: Synchronize::new(execution.max_threads),
+            });
+
+            RwLock { obj }
+        })
+    }
+
+    fn get_state<'a>(&self, objects: &'a mut object::Store) -> &'a mut State {
```

Comment:
Nit/style: I'd prefer it if we organized this impl block with all the `pub(crate)` methods first, followed by the private methods...not going to block merging the PR on that, though.

Reply:
Shouldn't it be called `state` according to API guidelines?

Reply:
An identical method is called `get_state` in Mutex FWIW.

Reply:
@pickfire I also thought about mentioning that, but I saw that `Mutex` had a function called `get_state`, and I thought consistency was good... 🤷‍♀